test/spl/degree_elevation_test.cc

```diff
@@ -0,0 +1,305 @@
+    std::vector<baf::ControlPoint> control_points = {
+        baf::ControlPoint(std::vector<double>({1.0, 1.0})), baf::ControlPoint(std::vector<double>({1.0, 2.0})),
+        baf::ControlPoint(std::vector<double>({2.0, 2.0})), baf::ControlPoint(std::vector<double>({2.0, 3.0})),
+        baf::ControlPoint(std::vector<double>({3.0, 3.0})), baf::ControlPoint(std::vector<double>({3.0, 4.0})),
+
+        baf::ControlPoint(std::vector<double>({2.0, 1.0})), baf::ControlPoint(std::vector<double>({1.5, 2.5})),
+        baf::ControlPoint(std::vector<double>({2.0, 3.0})), baf::ControlPoint(std::vector<double>({2.5, 3.5})),
+        baf::ControlPoint(std::vector<double>({3.5, 4.0})), baf::ControlPoint(std::vector<double>({5.0, 5.0}))
+    };
+    original_ = std::make_shared<spl::BSpline<2>>(knot_vector_before, degree, control_points);
+    elevated_ = std::make_shared<spl::BSpline<2>>(*original_);
+    elevated_->ElevateDegreeForDimension(0);
+  }
+
+ protected:
+  std::shared_ptr<spl::BSpline<2>> original_;
+  std::shared_ptr<spl::BSpline<2>> elevated_;
+};
+
+TEST_F(A2DBSplineForDegreeElevationForDimension0, HasELevatedDegreeInDimension0) {  // NOLINT
+  ASSERT_THAT(elevated_->GetDegree(0).get(), original_->GetDegree(0).get() + 1);
+  ASSERT_THAT(elevated_->GetDegree(1).get(), original_->GetDegree(1).get());
+}
+
+TEST_F(A2DBSplineForDegreeElevationForDimension0, HasNumberOfDifferentKnotsMoreKnots) {  // NOLINT
+  ASSERT_THAT(elevated_->GetKnotVector(0)->GetNumberOfDifferentKnots(),
+              original_->GetKnotVector(0)->GetNumberOfDifferentKnots());
+  ASSERT_THAT(elevated_->GetKnotVector(0)->GetNumberOfKnots(),
+              original_->GetKnotVector(0)->GetNumberOfKnots() +
+                  original_->GetKnotVector(0)->GetNumberOfDifferentKnots());
+  ASSERT_THAT(elevated_->GetKnotVector(1)->GetNumberOfKnots(), original_->GetKnotVector(1)->GetNumberOfKnots());
+}
+
+TEST_F(A2DBSplineForDegreeElevationForDimension0, HasNumberOfNonZeroKnotSpansMoreControlPoints) {  // NOLINT
+  ASSERT_THAT(elevated_->GetPointsPerDirection()[0],
+              original_->GetPointsPerDirection()[0] + original_->GetKnotVector(0)->GetNumberOfDifferentKnots() - 1);
+  ASSERT_THAT(elevated_->GetPointsPerDirection()[1], original_->GetPointsPerDirection()[1]);
+}
+
+TEST_F(A2DBSplineForDegreeElevationForDimension0, DoesNotChangeGeometricallyAfterDegreeElevation) {  // NOLINT
+  ASSERT_THAT(elevated_->AreGeometricallyEqual(*original_), true);
+}
+
+TEST_F(A2DBSplineForDegreeElevationForDimension0, DoesNotChangeGeometricallyAfterMoreDegreeElevations) {  // NOLINT
+  elevated_->ElevateDegreeForDimension(1);
+  ASSERT_THAT(elevated_->AreGeometricallyEqual(*original_), true);
+  elevated_->ElevateDegreeForDimension(0);
+  ASSERT_THAT(elevated_->AreGeometricallyEqual(*original_), true);
+  elevated_->ElevateDegreeForDimension(1);
+  ASSERT_THAT(elevated_->AreGeometricallyEqual(*original_, util::NumericSettings<double>::kEpsilon() * 5.0), true);
```

Comment:
We should be consistent in choosing the tolerance. I normally used 1e-10 if util::NumericSettings<double>::kEpsilon() didn't work. I don't know which solution is better but it should be done consistent for these test cases.